### ISSUE TITLE
Execution of trainer/train.py stops in the middle.

### ISSUE THREAD

**Issue by shu1106** [google-research/deeplab2#54]
I executed the following command, but it stopped in the middle of execution.

>python trainer/train.py \
>     --config_file=configs/coco/max_deeplab/max_deeplab_s_os16_res641_400k.textproto \
>     --mode=train \
>     --model_dir=checkpoints/max_deeplab_s_os16_res641_400k_coco_train \
>     --num_gpus=1

- execution results
----------------------------------------
>2021-09-15 05:26:54.602208: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic >library libcudart.so.11.0
>I0915 05:26:56.178199 139653727340352 train.py:67] Reading the config file.
I0915 05:26:56.181392 139653727340352 train.py:71] Starting the experiment.
2021-09-15 05:26:56.182713: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcuda.so.1
2021-09-15 05:26:56.223570: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.224579: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1734] Found device 0 with properties: 
pciBusID: 0000:00:1e.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-09-15 05:26:56.224627: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcudart.so.11.0
2021-09-15 05:26:56.229769: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcublas.so.11
2021-09-15 05:26:56.229851: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcublasLt.so.11
2021-09-15 05:26:56.232521: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcufft.so.10
2021-09-15 05:26:56.232933: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcurand.so.10
2021-09-15 05:26:56.233980: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcutensor.so.1
2021-09-15 05:26:56.234695: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcusolver.so.11
2021-09-15 05:26:56.235845: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcusparse.so.11
2021-09-15 05:26:56.236159: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcudnn.so.8
2021-09-15 05:26:56.236387: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.237526: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.238509: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1872] Adding visible gpu devices: 0
2021-09-15 05:26:56.239783: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.240771: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1734] Found device 0 with properties: 
pciBusID: 0000:00:1e.0 name: Tesla V100-SXM2-16GB computeCapability: 7.0
coreClock: 1.53GHz coreCount: 80 deviceMemorySize: 15.78GiB deviceMemoryBandwidth: 836.37GiB/s
2021-09-15 05:26:56.241016: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.242099: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.243114: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1872] Adding visible gpu devices: 0
2021-09-15 05:26:56.243177: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcudart.so.11.0
2021-09-15 05:26:56.927719: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1258] Device interconnect StreamExecutor with strength 1 edge matrix:
2021-09-15 05:26:56.927789: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1264]      0 
2021-09-15 05:26:56.927823: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1277] 0:   N 
2021-09-15 05:26:56.928101: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.929215: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.930264: I tensorflow/stream_executor/cuda/cuda_gpu_executor.cc:1050] successful NUMA node read from SysFS had negative value (-1), but there must be at least one NUMA node, so returning NUMA node zero
2021-09-15 05:26:56.931224: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1418] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 14652 MB memory) -> physical GPU (device: 0, name: Tesla V100-SXM2-16GB, pci bus id: 0000:00:1e.0, compute capability: 7.0)
2021-09-15 05:26:56.931653: I tensorflow/core/common_runtime/process_util.cc:146] Creating new thread pool with default inter op setting: 8. Tune using inter_op_parallelism_threads for best performance.
I0915 05:26:56.932756 139653727340352 train_lib.py:107] Using strategy <class 'tensorflow.python.distribute.one_device_strategy.OneDeviceStrategy'> with 1 replicas
I0915 05:26:57.241827 139653727340352 deeplab.py:57] Synchronized Batchnorm is used.
I0915 05:26:57.242810 139653727340352 axial_resnet_instances.py:144] Axial-ResNet final config: {'num_blocks': [3, 4, 6, 3], 'backbone_layer_multiplier': 1.0, 'width_multiplier': 1.0, 'stem_width_multiplier': 1.0, 'output_stride': 16, 'classification_mode': False, 'backbone_type': 'resnet_beta', 'use_axial_beyond_stride': 16, 'backbone_use_transformer_beyond_stride': 32, 'extra_decoder_use_transformer_beyond_stride': 32, 'backbone_decoder_num_stacks': 0, 'backbone_decoder_blocks_per_stage': 1, 'extra_decoder_num_stacks': 0, 'extra_decoder_blocks_per_stage': 1, 'max_num_mask_slots': 128, 'num_mask_slots': 128, 'memory_channels': 256, 'base_transformer_expansion': 1.0, 'global_feed_forward_network_channels': 256, 'high_resolution_output_stride': 4, 'activation': 'relu', 'block_group_config': {'attention_bottleneck_expansion': 2, 'drop_path_keep_prob': 0.800000011920929, 'drop_path_beyond_stride': 16, 'drop_path_schedule': 'linear', 'positional_encoding_type': None, 'use_global_beyond_stride': 0, 'use_sac_beyond_stride': -1, 'use_squeeze_and_excite': False, 'conv_use_recompute_grad': False, 'axial_use_recompute_grad': True, 'recompute_within_stride': 0, 'transformer_use_recompute_grad': False, 'axial_layer_config': {'query_shape': (129, 129), 'key_expansion': 1, 'value_expansion': 2, 'memory_flange': (32, 32), 'double_global_attention': False, 'num_heads': 8, 'use_query_rpe_similarity': True, 'use_key_rpe_similarity': True, 'use_content_similarity': True, 'retrieve_value_rpe': True, 'retrieve_value_content': True, 'initialization_std_for_query_key_rpe': 1.0, 'initialization_std_for_value_rpe': 1.0, 'self_attention_activation': 'softmax'}, 'dual_path_transformer_layer_config': {'num_heads': 8, 'bottleneck_expansion': 2, 'key_expansion': 1, 'value_expansion': 2, 'feed_forward_network_channels': 2048, 'use_memory_self_attention': True, 'use_pixel2memory_feedback_attention': True, 'transformer_activation': 'softmax'}}, 'bn_layer': functools.partial(<class 'tensorflow.python.keras.layers.normalization_v2.SyncBatchNormalization'>, momentum=0.9900000095367432, epsilon=0.0010000000474974513), 'conv_kernel_weight_decay': 0.0}
I0915 05:26:57.593706 139653727340352 deeplab.py:96] Setting pooling size to (41, 41)
I0915 05:26:57.593983 139653727340352 aspp.py:135] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0915 05:27:03.713764 139653727340352 controller.py:391] restoring or initializing model...
restoring or initializing model...
I0915 05:27:03.750207 139653727340352 controller.py:395] restored model from checkpoints/max_deeplab_s_os16_res641_400k_coco_train/$Panoptic_Test/ckpt-0.
restored model from checkpoints/max_deeplab_s_os16_res641_400k_coco_train/$Panoptic_Test/ckpt-0.
I0915 05:27:03.750401 139653727340352 controller.py:217] restored from checkpoint: checkpoints/max_deeplab_s_os16_res641_400k_coco_train/$Panoptic_Test/ckpt-0
restored from checkpoint: checkpoints/max_deeplab_s_os16_res641_400k_coco_train/$Panoptic_Test/ckpt-0
I0915 05:27:04.950217 139653727340352 api.py:446] Eval with scales ListWrapper([1.0])
I0915 05:27:06.379980 139653727340352 api.py:446] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0915 05:27:06.411498 139653727340352 api.py:446] Eval scale 1.0; setting pooling size to [41, 41]
WARNING:tensorflow:From /usr/local/lib/python3.8/dist-packages/tensorflow/python/ops/array_ops.py:5043: calling gather (from tensorflow.python.ops.array_ops) with validate_indices is deprecated and will be removed in a future version.
Instructions for updating:
The `validate_indices` argument has no effect. Indices are always validated on CPU and never validated on GPU.
W0915 05:27:09.029129 139653727340352 deprecation.py:528] From /usr/local/lib/python3.8/dist-packages/tensorflow/python/ops/array_ops.py:5043: calling gather (from tensorflow.python.ops.array_ops) with validate_indices is deprecated and will be removed in a future version.
Instructions for updating:
The `validate_indices` argument has no effect. Indices are always validated on CPU and never validated on GPU.
I0915 05:27:17.629784 139653727340352 api.py:446] Global average pooling in the ASPP pooling layer was replaced with tiled average pooling using the provided pool_size. Please make sure this behavior is intended.
I0915 05:27:18.935303 139653727340352 controller.py:236] train | step:      0 | training until step 400000...
train | step:      0 | training until step 400000...
2021-09-15 05:27:19.169520: I tensorflow/compiler/mlir/mlir_graph_optimization_pass.cc:176] None of the MLIR Optimization Passes are enabled (registered 2)
2021-09-15 05:27:19.188923: I tensorflow/core/platform/profile_utils/cpu_utils.cc:114] CPU Frequency: 2300015000 Hz
WARNING:tensorflow:From /usr/local/lib/python3.8/dist-packages/tensorflow/python/util/dispatch.py:206: calling foldl_v2 (from tensorflow.python.ops.functional_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.foldl(fn, elems, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.foldl(fn, elems))
W0915 05:27:32.651105 139653727340352 deprecation.py:596] From /usr/local/lib/python3.8/dist-packages/tensorflow/python/util/dispatch.py:206: calling foldl_v2 (from tensorflow.python.ops.functional_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.foldl(fn, elems, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.foldl(fn, elems))
WARNING:tensorflow:From /usr/local/lib/python3.8/dist-packages/tensorflow/python/autograph/impl/api.py:463: calling while_loop_v2 (from tensorflow.python.ops.control_flow_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.while_loop(c, b, vars, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.while_loop(c, b, vars))
W0915 05:27:33.151738 139653727340352 deprecation.py:596] From /usr/local/lib/python3.8/dist-packages/tensorflow/python/autograph/impl/api.py:463: calling while_loop_v2 (from tensorflow.python.ops.control_flow_ops) with back_prop=False is deprecated and will be removed in a future version.
Instructions for updating:
back_prop=False is deprecated. Consider using tf.stop_gradient instead.
Instead of:
results = tf.while_loop(c, b, vars, back_prop=False)
Use:
results = tf.nest.map_structure(tf.stop_gradient, tf.while_loop(c, b, vars))
2021-09-15 05:28:59.687466: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcudnn.so.8
2021-09-15 05:29:00.411789: I tensorflow/stream_executor/cuda/cuda_dnn.cc:380] Loaded cuDNN version 8202
2021-09-15 05:29:01.223826: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcublas.so.11
2021-09-15 05:29:01.834288: I tensorflow/stream_executor/platform/default/dso_loader.cc:54] Successfully opened dynamic library libcublasLt.so.11`
--------------------------------------------------------------

The following commands in the installation worked well.

`python deeplab2/tensorflow_ops/python/kernel_tests/merge_semantic_and_instance_maps_op_test.py`
`python deeplab2/model/deeplab_test.py`
`python deeplab2/trainer/evaluator_test.py`

I also referred to problem #36 , but it did not work.
Please let me know the solution.

- environment
  - NVIDIA Driver Version: 470.57.02    CUDA Version: 11.4 
  - nvidia docker installed on AWS (p3.xlarge)
    The docker file I wrote is as follows.

--------------------------
>FROM nvcr.io/nvidia/tensorflow:21.08-tf2-py3
>
>RUN mkdir /projects
>WORKDIR /projects
>
>RUN  apt-get update \
>         && apt-get install -y git \
>         cmake \
>         protobuf-compiler \
>         && apt-get update \
>         && apt clean
>
>RUN pip install --upgrade pip
>
>RUN  pip install numpy \
>        && pip install Pillow \
>        && pip install matplotlib \
>        && pip install cython \
>        && pip install jupyter
>
>RUN git clone https://github.com/google-research/deeplab2.git
>
>RUN git clone https://github.com/tensorflow/models.git
>
>RUN git clone https://github.com/cocodataset/cocoapi.git \
>        && cd cocoapi/PythonAPI \
>        && python setup.py install \
>        && make
>
>ENV PYTHONPATH $PYTHONPATH:`pwd`:`pwd`/models:`pwd`/cocoapi/PythonAPI
>RUN echo $PYTHONPATH
>
>RUN protoc deeplab2/*.proto --python_out=.
--------------------------------------------------------

**Comment by csrhddlam** [comment]
Hi,

Thanks for your interest in our work. I think this issue is probably because [MaX-DeepLab](https://github.com/google-research/deeplab2/blob/main/g3doc/projects/max_deeplab.md) requires TensorFlow>=2.6.0, while [your docker tensorflow:21.08](https://docs.nvidia.com/deeplearning/frameworks/tensorflow-release-notes/rel_21-08.html#rel_21-08) uses TensorFlow==2.5.0.

For more information about why TensorFlow 2.5 does not work, please refer to issue #24.
For more elaborative error messages, you could also try disabling the loop in Orbit as is done in issue #24. The orbit loop suppresses error messages.

Best,
Huiyu

**Comment by aquariusjay** [comment]
Closing the issue since it is not active for a while.
Please feel free to reopen it if needed.
